lib/capistrano/configuration/scm_resolver.rb

```diff
@@ -0,0 +1,149 @@
+module Capistrano
+  class Configuration
+    # In earlier versions of Capistrano, users would specify the desired SCM
+    # implementation using `set :scm, :git`, for example. Capistrano would then
+    # load the matching .rb file based on this variable.
+    #
+    # Now we expect users to explicitly `require` and call `new` on the desired
+    # SCM implementation in their Capfile. The `set` technique is deprecated.
+    #
+    # This SCMResolver class takes care of managing the transition from the old
+    # to new system. It maintains the legacy behavior, but prints deprecation
+    # warnings when it is used.
+    #
+    # To maintain backwards compatibility, the resolver will load the Git SCM by
+    # if default it determines that no SCM has been explicitly specified or
+    # loaded. To force no SCM to be used at all, use `set :scm, nil`. This hack
+    # won't be necessary once backwards compatibility is removed in a future
+    # version.
+    #
+    # TODO: Remove this class entirely in Capistrano 4.0.
+    #
+    class SCMResolver
+      DEFAULT_GIT = :"default-git"
+
+      include Capistrano::DSL
+
+      def resolve
+        return if scm_name.nil?
+        set(:scm, :git) if using_default_scm?
+
+        print_deprecation_warnings_if_applicable
+        return if scm_plugin_loaded?
+
+
+        if built_in_scm_name?
+          load_built_in_scm
+        else
+          # Compatibility with existing 3.x third-party SCMs
+          register_legacy_scm_hooks
+          load_legacy_scm_by_name
+        end
+      end
+
+      private
+
+      def using_default_scm?
+        return @using_default_scm if defined? @using_default_scm
+        @using_default_scm = (fetch(:scm) == DEFAULT_GIT)
+      end
+
+      # This is somewhat of a hack, because there is no guarantee that a third-
+      # party SCM will necessarily implement the typical SCM tasks. But it works
+      # well enough for the built-in SCMs.
+      def scm_plugin_loaded?
+        Rake::Task.tasks.any? { |t| t.name =~ /^[^:]+:create_release/ }
```

Comment:
Would it make more sense to require that scm plugins register themselves to avoid this? This feels like it could lead to some difficult to debug situations for a developer creating a new scm plugin.